illumos-utils/src/zfs.rs

```diff
@@ -236,56 +237,123 @@ pub struct DatasetProperties {
 }
 
 impl DatasetProperties {
-    // care about.
-    const ZFS_LIST_STR: &'static str =
+    const ZFS_GET_PROPS: &'static str =
         "oxide:uuid,name,avail,used,quota,reservation,compression";
 }
 
-// An inner parsing function, so that the FromStr implementation can always emit
-// the string 's' that failed to parse in the error message.
-fn dataset_properties_parse(
-    s: &str,
-) -> Result<DatasetProperties, anyhow::Error> {
-    let mut iter = s.split_whitespace();
+impl DatasetProperties {
+    /// Parses dataset properties, assuming that the caller is providing the
+    /// output of the following command as stdout:
+    ///
+    /// zfs get -rpo name,property,value,source $ZFS_GET_PROPS $DATASETS
+    fn parse_many(
+        stdout: &str,
+    ) -> Result<Vec<DatasetProperties>, anyhow::Error> {
+        let name_prop_val_source_list = stdout.trim().split('\n');
 
-    let id = match iter.next().context("Missing UUID")? {
-        "-" => None,
-        anything_else => Some(anything_else.parse::<DatasetUuid>()?),
-    };
+        let mut datasets: BTreeMap<&str, BTreeMap<&str, _>> = BTreeMap::new();
+        for name_prop_val_source in name_prop_val_source_list {
+            let mut iter = name_prop_val_source.split_whitespace();
```

Comment:
I realize this particular bit didn't change in this PR, but splitting on whitespace seems a little fishy (although having to parse CLI output is probably going to be a little fishy no matter what). I was going to ask if columns could themselves have whitespace, but it looks like that's already true: the source isn't just `"inherited"`; it's really `"inherited from ..."`, and we just happen to be able to treat it as `inherited` because we're ignoring any extra data after the first space in the `source` column.

The man page says `-H` emits tab-delimited columns. Could we use that to make this more robust? Something like:

* Split on `\t` specifically instead of any whitespace
* Confirm there are no unexpected columns (return an error if `iter.next()` returns another string, I think?) after confirming the columns we do expect
* Fix up the places where we're relying on whitespace parsing to shorten the value (source would now start with `inherited from`; I don't see any other places but might be missing something not in the diff?)

Reply:
Yeah, seems totally reasonable to me. I also felt a little gross about this (tried to cover both cases in tests) but splitting on tab explicitly seems better.